src/main/java/io/neonbee/data/DataContext.java

```diff
@@ -96,6 +98,105 @@ public interface DataContext {
      */
     <T> T remove(String key);
 
+    /**
+     * Arbitrary response data of this context, which is passed through each processing step of the verticle backwards
+     * from callee to caller.
+     *
+     * @return all the context response data as a map.
+     */
+    Map<String, Object> responseData();
+
+    /**
+     * Sets all the response data.
+     *
+     * @param data the data, must be compatible to {@link JsonObject#JsonObject(Map)}
+     * @return a reference to this DataContext for chaining
+     */
+    DataContext setResponseData(Map<String, Object> data);
+
+    /**
+     * Merges the response data of a given map into the existing response context data map.
+     * <p>
+     * Note: This method does not perform a deep merge operation, but overrides already existing elements w/o merging
+     *
+     * @param data the response data to merge, must be compatible to {@link JsonObject#JsonObject(Map)}
+     * @return a reference to this DataContext for chaining
+     */
+    DataContext mergeResponseData(Map<String, Object> data);
+
+    /**
+     * Put some arbitrary response data in the context. This will be available in any data verticle that receive the
+     * context.
+     *
+     * @param key   the key for the data
+     * @param value the data, must be compatible to {@link JsonObject#put(String, Object)} and should be primitive,
+     *              otherwise it can't be serialized via the event bus by default.
+     * @return a reference to this DataContext for chaining
+     */
+    DataContext putResponseDataEntry(String key, Object value);
+
+    /**
+     * Get some response data from the context. The data is available in any data verticle that receive the context.
+     *
+     * @param key the key for the data
+     * @param <T> the type of the parameter to be returned
+     * @return the data, as expected from {@link JsonObject#getValue(String)}
+     */
+    <T> T getResponseDataEntry(String key);
+
+    /**
+     * Remove some response data from the context. The data is available in any data verticle that receive the context.
+     *
+     * @param key the key for the data
+     * @param <T> the type of the parameter to be returned
+     * @return the previous data associated with the key, as expected from {@link JsonObject#remove(String)}
+     */
+    <T> T removeResponseDataEntry(String key);
+
+    /**
+     * Returns a map between {@link DataRequest} and received data map.
+     *
+     * @return a map between {@link DataRequest} and received data map
+     */
+    Map<DataRequest, Map<String, Object>> receivedData();
+
+    /**
+     * Sets the response data map for all {@link DataRequest}s.
+     *
+     * @param map a map between {@link DataRequest} and received response data for the request.
+     * @return a reference to this DataContext for chaining
+     */
+    DataContext setReceivedData(Map<DataRequest, Map<String, Object>> map);
```

Comment:
this should never be possible to be called by an end-user, so please remove

Reply:
A setter is but required to set the received data in DataVerticle (RecursiveResolutionRoutine) into the DataContext.

Since DataContextImpl is not in the same package as DataVerticle, a package protected method will not work.